### PR TITLE
Enable strict mypy type checking 

### LINKED ISSUE #88: Mypy does not check types in test code
**What is the expected feature or enhancement?**
Mypy should also check test code and fail if typing issues are found. Would have avoided https://github.com/Qiskit/qiskit-ibm-runtime/pull/80#pullrequestreview-848478338 where tests had to fail to uncover that tests have not been updated to the latest method signatures.

### PR DESCRIPTION
**Pull request by kt474**
### Summary

I assume the purpose of #88 is to enable `show_none_errors` which will shows errors related to strict `None` checking - not sure if this would've caught the issue [here](https://github.com/Qiskit/qiskit-ibm-runtime/pull/80#pullrequestreview-848478338). Everything else in mypy is already enabled.

Enabling `show_none_errors` gives 52 "errors" - I'll have to go through each of them but there will be merge conflicts with the current PRs and I'm not sure all of the errors are worth fixing

### Details and comments
Fixes #88

### CODE REVIEW

**Comment by daka1510** [comment]
> I assume the purpose of https://github.com/Qiskit/qiskit-ibm-runtime/issues/88 is to enable show_none_errors which will shows errors related to strict None checking - not sure if this would've caught the issue https://github.com/Qiskit/qiskit-ibm-runtime/pull/80#pullrequestreview-848478338. Everything else in mypy is already enabled.

No, the intent of https://github.com/Qiskit/qiskit-ibm-runtime/issues/88 is to ensure that `mypy` also checks for type errors in our test code. The command we currently use does not run mypy for all files in the `test` directory. 
```
mypy:
	mypy --module qiskit_ibm_runtime
```

File: setup.cfg [hunk 1]
```diff
@@ -15,3 +16,6 @@ strict_optional = False
 show_none_errors = False
 show_error_codes = True
 no_site_packages = True
+
+[mypy-test.*]
+ignore_errors = True
```
Comment (kt474) [inline setup.cfg:21]: Which specific rules do we want to enable here? When enabled, there are over 400 errors - mostly `no-untyped-def`, and `no-untyped-call`

Also, if `test_jupyter` is not excluded, there are a lot errors in `qiskit_ibm_runtime/visualization` and `qiskit_ibm_runtime/jupyter`

Comment (daka1510) [reply setup.cfg:21]: `qiskit_ibm_runtime/visualization` and `qiskit_ibm_runtime/jupyter` are dead code paths that we will follow up on via https://github.com/Qiskit/qiskit-ibm-runtime/issues/112, https://github.com/Qiskit/qiskit-ibm-runtime/issues/124. So I agree to exclude it from type checking.

I suggest to start with
```
[mypy-test.*]
disallow_untyped_calls = False
disallow_untyped_defs = False
```
and investigate/fix all remaining 60 issues. 41 issues can be fixed very easily by adding an explicit return type (e.g. `-> None`).
```
# appears like disallow_untyped_defs does not apply here. If function parameters are partially typed, mypy expects the return type to be explicitly set
test/integration/test_proxies.py:120: error: Function is missing a return type annotation  [no-untyped-def]
```

File: test/ibm_test_case.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+# mypy: ignore-errors
```
Comment (daka1510) [inline test/ibm_test_case.py:1]: Why is this needed?

File: test/unit/mock/fake_runtime_service.py [hunk 3]
```diff
@@ -52,11 +52,11 @@ def __init__(self, *args, **kwargs):
         ):
             super().__init__(*args, **kwargs)
 
-    def _authenticate_legacy_account(self, client_params: ClientParameters):
+    def _authenticate_legacy_account(self, client_params: ClientParameters) -> Any:
```
Comment (daka1510) [inline test/unit/mock/fake_runtime_service.py:55]: Does `-> FakeAuthClient` not work?

**Review by daka1510 — COMMENTED** [review]
Thanks for following up, @kt474. Added 2 questions, otherwise looks good.

File: test/unit/mock/fake_runtime_service.py (hunk 3 above)
Comment (kt474) [reply test/unit/mock/fake_runtime_service.py:55]:
```
Return type "FakeAuthClient" of "_authenticate_legacy_account" incompatible with return type "AuthClient" in supertype "IBMRuntimeService"
```

There is a mypy override error if `FakeAuthClient` is used as a return type so we can have `Any` as a return type or do something like ```-> 'FakeAuthClient' # type: ignore```

File: test/ibm_test_case.py (hunk 2 above)
Comment (kt474) [reply test/ibm_test_case.py]: There are `attr-defined` errors in the setup and teardown class methods - I can ignore the specific lines instead of ignoring errors for the whole file

Comment (daka1510) [reply test/ibm_test_case.py]: Can we also try to fix it in the test code? Would something like below work instead?

```
class IBMTestCase(unittest.TestCase):
    """Custom TestCase for use with the Qiskit IBM Runtime."""

    log: logging.Logger
    dependencies: IntegrationTestDependencies
    service: IBMRuntimeService
```

Comment (daka1510) [reply test/ibm_test_case.py]: That way we can benefit from auto-completion in subclasses that access the class attributes :)

image

**Review by daka1510 — APPROVED** [review]
Thanks for following up on my proposals, @kt474. Looks good to me. @rathishcholarajan any objections from your side to integrate it?

**Review by rathishcholarajan — APPROVED** [review]
LGTM!
